features/features.go

```diff
@@ -0,0 +1,124 @@
+type SnapdFeature int
+
+const (
+	// Layouts controls availability of snap layouts.
+	Layouts SnapdFeature = iota
+	// ParallelInstances controls availability installing a snap multiple times.
+	ParallelInstances
+	// Hotplug controls availability of dynamically creating slots based on system hardware.
+	Hotplug
+	// SnapdSnap controls possibility of installing the snapd snap.
+	SnapdSnap
+	// PerUserMountNamespace controls the persistence of per-user mount namespaces.
+	PerUserMountNamespace
+	// lastFeature is the final known feature, it is only used for testing.
+	lastFeature
+)
+
+// KnownFeatures returns the list of all known features.
+func KnownFeatures() []SnapdFeature {
+	return []SnapdFeature{Layouts, ParallelInstances, Hotplug, SnapdSnap, PerUserMountNamespace}
+}
+
+// featyreNames maps feature constant to stable string representation.
+// The constants here must be synchronized with cmd/libsnap-confine-private/feature.c
+var featureNames = map[SnapdFeature]string{
+	Layouts:               "layouts",
+	ParallelInstances:     "parallel-instances",
+	Hotplug:               "hotplug",
+	SnapdSnap:             "snapd-snap",
+	PerUserMountNamespace: "per-user-mount-namespace",
+}
+
+// String returns the name of a snapd feature.
+// The function panics for bogus feature values.
+func (f SnapdFeature) String() string {
+	if name, ok := featureNames[f]; ok {
+		return name
+	}
+	panic(fmt.Sprintf("unknown feature flag code %d", f))
+}
+
+// IsEnabledWhenUnset returns true if a feature is enabled when not set.
+//
+// A feature may be enabled or disabled with explicit state in snapd. If
+// explicit state is absent the effective value is the implicit default
+// computed by this function.
+func (f SnapdFeature) IsEnabledWhenUnset() bool {
+	switch f {
+	case Layouts:
+		return true
+	}
+	return false
+}
+
+// IsExported returns true if a feature is copied from snapd state to a feature file.
+//
+// Certain features are available outside of snapd internal state and visible as control
+// files in a dedicated directory. Such features can be queried for, via IsEnabled, outside
+// of snapd.
+func (f SnapdFeature) IsExported() bool {
+	switch f {
+	case PerUserMountNamespace:
+		return true
+	}
+	return false
+}
+
+// ControlFile returns the path of the file controlling the exported feature.
+//
+// Snapd considers the feature enabled if the file is present.
+// The contents of the file are not important.
+//
+// The function panics for features that are not exported.
+func (f SnapdFeature) ControlFile() string {
+	if !f.IsExported() {
+		panic(fmt.Sprintf("cannot compute the control file of feature %q because that feature is not exported", f))
+	}
+	return filepath.Join(dirs.FeaturesDir, f.String())
+}
+
+// ConfigOption returns the snap name and configuration option associated with this feature.
+func (f SnapdFeature) ConfigOption() (snapName, confName string) {
+	return "core", "experimental." + f.String()
```

Comment:
can we think of features that would not be just system options but on a different snap?  I sort think it would be better if the hack of attaching options to the "core" snap name even on core18 didn't live here as well, so this just returned confName.

Reply:
I was thinking about this and for now I don't have any ideas. I chose to put the snap name here since the consumers of this API have to provide one. It felt odd to hard-code one decision there and one decision here.

Reply:
I was thinking about this and perhaps we will do this kind of processing with gadget options. Those are not features yet but then again, who knows. Perhaps secure boot?

Reply:
Don't think we have clear use case yet for anything but system options. I have a slight preference for not mentioning "core" here.